slides: a sliding window range iterator

wilzbach opened the pull request:
As you can see - I was a bit productive during this weekend ;-)

I propose here to add a `pairwise` range iterator that iterates over all k-mers of a range, e.g.

```
[0,1,2].slides(2).array == [[0, 1], [1,2]];
```
## Why is this useful?

For me it's kind of common to work with k-mers and texts and thus it's nice to have a handy iterator for this. See this simple example for how easy counting will get:

```
int[dstring] e;
"AGAGA"d.slides.each!(a => e[a]++);
assert(e == ["AG"d: 2, "GA"d: 2]);
```

and of course works like this for any k:

```
int[dstring] e;
"AGAGA"d.slides(3).each!(a => e[a]++);
assert(e == ["AGA"d: 2, "GAG"d: 1]);
```
## Current alternatives

It's possible to do this with `zip`, but it looks a bit ugly and is only practical for `n==2`.

```
auto arr = "AGAGA".array;
arr.zip(arr.save.dropOne).map!"a.expand.only".map!(to!string).each!(a => d[a]++);
```
## Trivia
- I think `pairs` is a fine name to, but that might conflict with other functions e.g. `byPairs`
- the default is set to 2, as that's the most common use case
- the implementation is based on `chunks`, but I couldn't figure out a proper way to copy/inherit this functionality and most functions are actually changed

edit: changed name to `slides`

yebblies commented:
Another way to say it is that it iterates over all sub-strings with a certain length.

greenify commented:
@yebblies you might be right that the documentation and name could be more intuitive. I looked at different standard libs and how they name this function.

tl;dr: `slides` seem the most intuitive choice.
Alternatively there is also `windows` or `cons` - has anyone an opinion one this?

Scala: sliding
-----------------

> An iterator that yields a sliding fixed-sized window of elements in this collection.

[sliding](http://www.scala-lang.org/docu/files/collections-api/collections_4.html)

Ruby: each_cons
----------------------

> Iterates the given block for each array of consecutive <n> elements. If no block is given, returns an enumerator.

[each_cons](http://ruby-doc.org/core-2.3.0/Enumerable.html#method-i-each_cons)

Clojure: partition
---------------------

> Usage: (partition n step pad coll)

> Returns a lazy sequence of lists of n items each, at offsets step
> apart. If step is not supplied, defaults to n, i.e. the partitions
> do not overlap. If a pad collection is supplied, use its elements as
> necessary to complete last partition upto n items. In case there are
> not enough padding elements, return a partition with less than n items.

[partition](https://clojure.github.io/clojure/clojure.core-api.html#clojure.core/partition)

quickfur commented:
The use of "pair" here, both in the function name and in the documentation, is very confusing, because "pair" means _two_ of something, but you're actually talking about k-mers, for any k≥2. (On that note, does the code handle the k=1 case nicely?)

IMO a better name would be something along the lines of a sliding window of size k, or something along those lines.

greenify commented:
> The use of "pair" here, both in the function name and in the documentation, is very confusing, because "pair" means two of something, but you're actually talking about k-mers, for any k≥2.
> IMO a better name would be something along the lines of a sliding window of size k, or something along those lines.

I renamed it to `slides` and updated the documentation to go with the term `fixed-size sliding window`.

> On that note, does the code handle the k=1 case nicely?

Yes - check the tests ;-)

burner commented on std/range/package.d:
```diff
@@ -6730,6 +6734,384 @@ if (isForwardRange!Source && hasLength!Source)
+    import std.array: array;
+    import std.algorithm: equal, map;
+    import std.conv: to;
+    assert(iota(3).slides(3).array.to!string == "[[0, 1, 2]]");
```
there is no reason to convert this to string just compare with [[0,1,2]] or use the algorithm function equal

greenify replied:
I didn't know that its as easy as using `equal!equal([[0,1,2]]))` - very nice :)

burner replied:
D is like vim. You learn something new every day.

greenify commented:
FYI: This is also part of ndslice.
http://dlang.org/phobos/std_experimental_ndslice_selection.html#.windows

greenify commented:
This also tackles parts of issue 6621
https://issues.dlang.org/show_bug.cgi?id=6621

JackStouffer commented on std/range/package.d:
```diff
@@ -6730,6 +6734,382 @@ if (isForwardRange!Source && hasLength!Source)
+        }
+    }
+
+private:
```
Two nitpicks:

1. Using Python's notation for private symbols with an underscore is useless with an explicit `private` feature
2. Struct members should be at the top

greenify replied:
> Using Python's notation for private symbols with an underscore is useless with an explicit private feature

I know, but this style keeps popping up in phobos and the method that `slides` is based on (`chunks`) used the same notation.

> Struct members should be at the top

Again kept the style from `chunks`, but I do agree that it should be at top - updated :)

greenify commented:
Anything else that is blocking this PR?  - I do believe that this is a quite common method and useful for phobos.

JakobOvrum commented on std/range/package.d:
```diff
@@ -6730,6 +6734,381 @@ if (isForwardRange!Source && hasLength!Source)
+of size $(D windowSize) of a $(D source) range.
+$(D Source) must be a forward range. $(D windowSize) must be
+greater than zero. For $(D windowSize = 2) it is similar to
+$D(zip(arr, arr.save.dropOne)).
```
`zip(source, source.save.dropOne)`?

JakobOvrum commented on std/range/package.d:
```diff
@@ -6730,6 +6734,381 @@ if (isForwardRange!Source && hasLength!Source)
+greater than zero. For $(D windowSize = 2) it is similar to
+$D(zip(arr, arr.save.dropOne)).
+
+$(D[1,2,3].slides().array == [[1,2], [2,3]])
```
This looks kind of out of place in the output and it's already in the example section, so maybe just remove it?

JakobOvrum commented on std/range/package.d:
```diff
@@ -6730,6 +6734,381 @@ if (isForwardRange!Source && hasLength!Source)
+
+See_Also: $(LREF chunks)
+
+Returns: Input range of all sliding windows
```
It's always at least a forward range with propagated bidirectionality and conditional random access and slicing, so stating "Input range" here is confusing. I'd prefer if all the capabilities were succinctly documented.

JakobOvrum commented on std/range/package.d:
```diff
@@ -6730,6 +6734,381 @@ if (isForwardRange!Source && hasLength!Source)
+
+Returns: Input range of all sliding windows
+*/
+struct Slides(Source)
```
We generally use nested structs for new higher order ranges, and we generally don't document range primitives individually.

Range primitives always have the same or very similar signature when present, and their presence is often conditional, which is not reflected in the documentation in any special way. I think the signal to noise ratio of higher order range documentation is much better when range capabilities are documented in prose in the main documentation comment.

JakobOvrum commented on std/range/package.d:
```diff
@@ -6730,6 +6734,381 @@ if (isForwardRange!Source && hasLength!Source)
+    {
+        assert(!_source.empty);
+        _source.popFront();
+        if(!_endSource.empty)
```
This check looks to be redundant

rcorre commented:
A bit late to the party, but I've needed this more than once so it would be great to have in phobos.
At the risk of bikeshedding, could it just be called `slide`? `chunks` is the only other plural name and it doesn't seem to fit here.

Just for fun, a compile-time windows size gives a succint definition and allows multiple foreach arguments:

http://dpaste.dzfl.pl/fced4f16a987

I haven't figured out a clever way to use that same technique for a runtime `windowSize` though.

greenify commented:
> We generally use nested structs for new higher order ranges, and we generally don't document range primitives individually.
> Range primitives always have the same or very similar signature when present, and their presence is often conditional, which is not reflected in the documentation in any special way. I think the signal to noise ratio of higher order range documentation is much better when range capabilities are documented in prose in the main documentation comment.

I fully agree with this, however when I looked at `package.d` I saw that most of the functions in this file - even similar ones like `chunks`, `evenChunks use a public struct. Are there any plan to move the structs into the function?
(moved the `slides` inside the function and addressed all comments)

JakobOvrum commented:
Those are quite old. I'm not sure if they were designed that way because of buggy/incomplete compilers or simply because practice regarding ranges hadn't yet fully evolved. Maybe a mix of both (and maybe there will be someone who comes out of the woodwork in support of the old practice…).

Regardless, changing them would require deprecating the public range types and reworking their documentation, and YMMV whether the latter is enough to justify the former. At any rate I don't know of any plan for them.

greenify commented:
> A bit late to the party, but I've needed this more than once so it would be great to have in phobos.
> At the risk of bikeshedding, could it just be called slide? chunks is the only other plural name and it doesn't seem to fit here.

ping @andralex - what is your opinion on this addition? `slides` or `slide`?

9il commented:
Please add additional variant, which can work with simple input ranges. This function should be included along with circular buffer range, thought. In addition, two kinds of sliding windows should be supported: one with fixed width and one with max length (less length for borders).

schveiguy commented:
I think we should not be adding more inner types, because they add significant template bloat. I experienced this to a huge degree when writing my [iopipe](https://github.com/schveiguy/iopipe) library.

See this forum post: https://forum.dlang.org/post/n96k3g$ka5$1@digitalmars.com

Note: if it is decided to keep the struct inside the function, there is no need to add the template parameter to the inner struct. It's already templated based on the function.

JackStouffer commented:
@wilzbach Seems like you could add a more performant version if the original range supports slicing that just returns slices for `front`.

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+Returns: Forward range of all sliding windows with propagated bidirectionality,
+         conditional random access and slicing.
+*/
+auto slides(Source)(Source source, size_t windowSize = 2)
```
I'm against the default windowSize. If you use slides you properly have a specific windowSize in mind. It just makes for more documentation and more reading.

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+            _endSource.popFront();
+        }
+
+        static if (!isInfinite!Source)
```
add curly braces to the static if and else as in the following static if

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+                immutable start = max(0, index);
+                immutable end   = start + _windowSize;
+
+                static if (isInfinite!Source)
```
please no mixed brace style

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+            }
+
+            /// Ditto
+            static if (hasLength!Source)
```
see comment above

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+                        return mom.length;
+                    }
+                    alias momLength this;
+                }
```
add empty line

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+                //These overloads avoid that.
+                typeof(this) opSlice(DollarToken, DollarToken)
+                {
+                    static if (hasSliceToEnd)
```
see comment above

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+                {
+                    import std.algorithm : min;
+                    assert(lower <= length, "slides slicing index out of bounds");
+                    static if (hasSliceToEnd)
```
see comment above

burner commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+                        return slides(_source[min(lower, len) .. len], _windowSize);
+                    }
+                }
+                typeof(this) opSlice(DollarToken, size_t upper)
```
add empty line

9il commented on std/range/package.d:
```diff
@@ -6777,6 +6790,376 @@ if (isForwardRange!Source && hasLength!Source)
+            void popBack()
+            {
+                assert(!empty, "popBack() called on empty slides");
+                immutable end = (_source.length - 1) / _windowSize * _windowSize;
```
where `end` is used?

9il commented:
@wilzbach Could you please provide a real world example where `windows` can not be used instead of `slides`? I would like to add `slides` but just for input ranges and with internal buffer.

wilzbach commented:
So I found a bit of time to rework this PR:

- for `InputRange` a simple buffer is used
- Flag to allow to change the behavior if the range has less elements than `windowSize`
- removed default value for `windowSize`
- added an optional parameter `stepSize` (default 1)
- fixed a couple of bugs and added a lot more tests :)
- proper capability propagation (a la Design by Introspection)
- addressed all other comments
- this now fixes 6621 completely

> Please add additional variant, which can work with simple input ranges. This function should be included along with circular buffer range, thought. In addition, two kinds of sliding windows should be supported: one with fixed width and one with max length (less length for borders).

Done. I set the visibility of the buffer to `private` for now.

> I think we should not be adding more inner types, because they add significant template bloat. I experienced this to a huge degree when writing my iopipe library.

Ok. done :)

> @wilzbach Seems like you could add a more performant version if the original range supports slicing that just returns slices for front.

Good idea. Done.

> @wilzbach Could you please provide a real world example where windows can not be used instead of slides? I would like to add slides but just for input ranges and with internal buffer.

AFAICT `windows` from ndslice only works if one knows the length and can thus use `sliced`. The range-based version here works with any `ForwardRange` and now even with any `InputRange`.

JackStouffer commented on std/range/package.d:
```diff
@@ -7232,6 +7245,953 @@ if (isForwardRange!Source && hasLength!Source)
+/**
+This range iterates over a fixed-sized sliding window (k-mers)
+of size `windowSize` of a `source` range.
+$(D Source) must be at least an `InputRange` and
```
>$(D Source)

backticks

JackStouffer commented on std/range/package.d:
```diff
@@ -7232,6 +7245,953 @@ if (isForwardRange!Source && hasLength!Source)
+For `windowSize = 1` it splits the range into one element groups.
+For `windowSize = 2` it is similar to `zip(source, source.save.dropOne)`.
+
+If the Range is a mere `InputRange` (it doesn't support saving its state
```
I think you can drop the input vs forward range explaination

JackStouffer commented on std/range/package.d:
```diff
@@ -7232,6 +7245,953 @@ if (isForwardRange!Source && hasLength!Source)
+    /// Forward range primitives. Always present.
+    @property auto front()
+    {
+        assert(!empty);
```
needs an error message

JackStouffer commented on std/range/package.d:
```diff
@@ -7232,6 +7245,953 @@ if (isForwardRange!Source && hasLength!Source)
+    /// Ditto
+    void popFront()
+    {
+        assert(!empty);
```
ditto

JackStouffer commented on std/range/package.d:
```diff
@@ -7232,6 +7245,953 @@ if (isForwardRange!Source && hasLength!Source)
+                import std.algorithm.comparison : min;
+
+                immutable start = index * _stepSize;
+                immutable end   = start + _windowSize;
```
Please don't line up equal signs

wilzbach commented:
> Apart from the little nitpicks, I'm skeptical about adding an allocating method to a range function. This is really outside the norm; usually if the user has an input range and wants to use it on a forward range requiring function, we ask them to make it an array.

The idea to use a CircularBuffer for input ranges was suggested by @9il - I like the idea that one needs to allocate a lot less than needed. Consider e.g. this example - we only need to allocate a small amount in the beginning:

```
int k;
generate!(_ => k++).slides(3, 7).filter!(x => x % 5 == 0).take(5).writeln;
// [[14, 15, 16], [49, 50, 51], [84, 85, 86], [119, 120, 121], [154, 155, 156]]
```

We could also make the `stepSize` a compile-time option?
This would allow the user to go for zero GC allocation even with mere input ranges.

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+    static if (hasLength!Source)
+    {
+        /// Length. Only if $(D hasLength!Source) is $(D true)
+        @property size_t length()
```
const

andralex replied:
drop `@property` :)

JackStouffer replied:
I've never been able to get `const` to work with ranges, and I doubt it will work here. It's because every range function inevitably calls a non-`const` function.

For example, `_source.length` may in fact be mutable for the underlying range.

andralex replied:
Ah, sorry I missed that it uses source.length. Never mind.

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+            {
+                import std.math : ceil;
+                double t = _source.length - _windowSize + 1;
+                return cast(size_t) ceil(t / _stepSize);
```
nah, that's overkill - use `(_source.length - _windowSize + _stepSize) / _stepSize`

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+            auto opIndex(size_t index)
+            {
+                import std.algorithm.comparison : min;
+
```
nospace

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+            }
+        }
+        // hasSlicing implies either isInfinite or hasLength
+        //else static if (hasSliceToEnd)
```
delete?

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+                // notice that we only need to move for windowSize - 1 to the right
+                // [0, 1, 2, 3] - slides(2) -> [[0, 1], [1, 2], [2, 3]]
+                // [0, 1, 2, 3] - slides(3) -> [[0, 1, 2], [1, 2, 3]]
+                return typeof(this)(_source[min(lower, len) .. min(upper + _windowSize - 1, len)], _windowSize, _stepSize);
```
long line even github dun like

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+                    return DollarToken();
+                }
+                //Slice to dollar
+                typeof(this) opSlice(size_t lower, DollarToken)
```
elegantly done

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+                    return mom.length;
+                }
+                alias momLength this;
+            }
```
this may be getting a little too pedantic

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+        }
+        else
+        {
+            alias ThisType = typeof(this);
```
private (also: why?)

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+            //operations such as : slides[$/2, $-1];
+            private static struct DollarToken
+            {
+                ThisType* mom;
```
no need to save the pointer, just the length (can't change for the lifetime of the token)

andralex commented on std/range/package.d:
```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+                assert(upper == length, "slides slicing index out of bounds");
+                return this[$ .. $];
+            }
+        }
```
looks like something went too far somewhere with this dollar business... not sure what use cases we're improving

JackStouffer commented:
Ping @wilzbach

JackStouffer commented:
I just found another use for this in my own code. 

```d
// lazily get all possible permutations from an array

auto a = ["Jack", "Joe", "John", "Jesus"];
auto app = appender!(string[]);

foreach (i; 2 .. a.length)
{
    put(
        a.sildes(i).permutations.joiner,
        app
    );
}
```

@wilzbach please consider addressing the comments here.

wilzbach commented:
Okay I managed to find a bit of time & went through `slides` again. In short I addressed the comments, added more tests, cleaned the code a bit more & added explanations for tricky bits (see e.g. `opSlice` or `back`)

> drop @property :)

Isn't this the typical range style to have `empty`, `front` `length` as `@property`?
(I know that `@property` for getter functions is nearly useless, hence "style")

> looks like something went too far somewhere with this dollar business... not sure what use cases we're improving

This was inspired by the way `chunk` handles the dollar tokens.

> Consider dropping bidirectional support. The overwhelming use will be forward.
> Sliding over input ranges is really a composition of "I need a buffered range that promotes my input range to a range with a lookahead" and, well, the existing code. I think we should leave this PR for the Slider for forward ranges, and then introduce Buffered with a new PR. Then the composition will be straightforward with little or no glue needed.

Alrighty -  I removed the buffered sliding window, but left the bidirectional support as it's very short code (once figured out correctly).

JackStouffer commented on std/range/package.d:
```diff
@@ -7299,6 +7314,978 @@ if (isForwardRange!Source && hasLength!Source)
+        enum needsEndTracker = true;
+    }
+
+    private bool _empty;
```
redundant `private`

JackStouffer commented on std/range/package.d:
```diff
@@ -7299,6 +7314,978 @@ if (isForwardRange!Source && hasLength!Source)
+
+    static if (hasSlicing!Source)
+    {
+        private enum hasSliceToEnd = hasSlicing!Source && is(typeof(Source.init[0 .. $]) == Source);
```
here too

JackStouffer commented on std/range/package.d:
```diff
@@ -7299,6 +7314,978 @@ if (isForwardRange!Source && hasLength!Source)
+                    {
+                        // if possible use the faster opDollar overload
+                        static if (hasSliceToEnd)
+                            _source = _source[$ .. $];
```
Often opDollar is just an alias to the length function. Why do you state that the `$` branch is faster?

JackStouffer commented on std/range/package.d:
```diff
@@ -7299,6 +7314,978 @@ if (isForwardRange!Source && hasLength!Source)
+    /// Forward range primitives. Always present.
+    @property auto front()
+    {
+        assert(!empty, "Attempting to access front on an empty range");
```
nitpick: 

```
Attempting to access front on an empty slides
```

Same for other error messages

JackStouffer commented on std/range/package.d:
```diff
@@ -7299,6 +7314,978 @@ if (isForwardRange!Source && hasLength!Source)
+    {
+        /**
+        Indexing and slicing operations. Provided only if
+        $(D hasSlicing!Source) is $(D true).
```
nitpick: backticks

JackStouffer commented on std/range/package.d:
```diff
@@ -7299,6 +7314,978 @@ if (isForwardRange!Source && hasLength!Source)
+
+        static if (!isInfinite!Source)
+        {
+            typeof(this) opSlice(size_t lower, size_t upper)
```
`/// ditto`

JackStouffer commented:
Ping @wilzbach 

this needs just a bit more work to be ready, let's get this merged

JackStouffer commented:
As @wilzbach seems to be busy I'll pull this and submit a follow up PR to address the nitpicks

wilzbach commented:
@ all: I just revived `std.range.slide` from the dead -> https://github.com/dlang/phobos/pull/5943 (and with a bit of luck it might be part of 2.079).
